### ISSUE TITLE
Add proxy/load-balancer/ssl termination support, like SAMLContextProviderLB

### ISSUE THREAD

**Issue by jdzijlstra** [spring-projects/spring-security#8977]
**Current Behavior**

I have a springboot application using SAML2 (using spring-security-saml2-service-provider). My application (the SP) needs to work with Microsoft Azure (the IDP) and is running in a docker image behind a proxy with SSL termination.
Azure (rightfully) requires my 'service-url' ( _assertionConsumerServiceUrlTemplate_) to use https.

This means that for Azure, my application is accessible via https only, where the proxy proxies to my application via http, so my application receives an http-request. Therefore my 'service-url' is set to an https-address, otherwise Azure cannot access my application via the proxy, but the application reconstructs the URL, via the request, to an http-address. 

The problem now is that the response from Azure is not accepted. I get the error: "An error occurred while validating the assertion: No subject confirmation methods were met for assertion with ID '_90d4b4e4-ba9f-44ee-9fa2-4cbf4f951f00'". 

Debug-logging shows

```
DEBUG [http-nio-9500-exec-7] org.opensaml.saml.saml2.assertion.impl.AbstractSubjectConfirmationValidator - Evaluating SubjectConfirmationData@Recipient of : https://myhost/login/saml2/sso/myapp
DEBUG [http-nio-9500-exec-7] org.opensaml.saml.saml2.assertion.impl.AbstractSubjectConfirmationValidator - Failed to match SubjectConfirmationData@Recipient to any supplied valid recipients: [http://myhost/login/saml2/sso/myapp]
DEBUG [http-nio-9500-exec-7] org.opensaml.saml.saml2.assertion.SAML20AssertionValidator - No subject confirmation methods were met for assertion with ID '_242557fc-f506-4c1d-b50a-49edd4430f00'
```

(of course _myhost_ and _myapp_ are replaced)

The SubjectConfirmationData@Recipient ('https://myhost/login/saml2/sso/myapp') is received from the IDP, but is actually the value set via 

        RelyingPartyRegistration.withRegistrationId(...) 
                                .localEntityIdTemplate(...) 
                                .assertionConsumerServiceUrlTemplate("https://myhost/login/saml2/sso/myapp")

The supplied valid recipients ('http://myhost/login/saml2/sso/myapp') is deducted from the http-request (in Saml2WebSsoAuth.attemptAuthentication: request.getRequestURL().toString()):

**Expected Behavior**

I'd like to be able to overrule the URL (valid recipients), and not always deduct it from the request.
In my case I would like to overrule the scheme, but I can image that also the hostname, port and path may be necessary to be overruled.

Actually, while thinking it over again, shouldn't the URL as received from the IDP not be compared to the value as configured in RelyingPartyRegistration.assertionConsumerServiceUrlTemplate(String)? So, we tell the IDP to contact us at a certain URL and in the message says it is meant for that URL and we check if that URL is indeed the URL we have configured.
Thus, i.s.o. passing request.getRequestURL().toString() to Saml2WebSsoAuth.attemptAuthentication, pass in the value based on rp.getAssertionConsumerServiceUrlTemplate() (and the request to compute {baseUrl} if it is in there)

**Context**

At this moment I don't see a workaround, other than let my application (Tomcat) handle the HTTPS itself, but this is not allowed by my company.
Another workaround is not to use spring-security-saml2-service-provider, but the 'old' spring-security-saml. This requires of course a rewrite of my code get that to work again.

**Comment by jdzijlstra** [comment]
Maybe this is a bug and not an enhancement request.

**Comment by jdzijlstra** [comment]
I've patched Saml2WebSsoAuthenticationFilter locally with

104c106
<                               request.getRequestURL().toString(),
---
>                               getAssertionConsumerServiceUrl(rp.getAssertionConsumerServiceUrlTemplate(), request, rp),
121c123,127
< }
---
>       private String getAssertionConsumerServiceUrl(final String applicationUri, final HttpServletRequest request, final RelyingPartyRegistration relyingParty)
>       {
>         final Function<String, String> resolver = templateResolver(applicationUri, relyingParty);
>         return resolver.apply(relyingParty.getAssertionConsumerServiceUrlTemplate());
>       }
122a129,133
>   private Function<String, String> templateResolver(String applicationUri, RelyingPartyRegistration relyingParty)
>   {
>     return template -> Saml2ServletUtils.resolveUrlTemplate(template, applicationUri, relyingParty);
>   }
> }

This works for me.

**Comment by jzheaux** [comment]
@jdzijlstra, I believe this is resolved by #8887, which was released as part of [`5.4.0-RC1`](https://github.com/spring-projects/spring-security/releases/tag/5.4.0-RC1). It ensures that the ACS URL is compared against `RelyingPartyRegistration#assertionConsumerServiceLocation`. Also, if that's not quite what you need, it provides an interface where you can modify how the `RelyingPartyRegistration` is resolved.

If after you try `5.4.0-RC1`, you discover that you need something more, please feel free to re-open the issue.
